### ISSUE TITLE
The fallback IdP requires a password

### ISSUE THREAD

**Issue by rbtcollins** [mozilla/persona#3494]
I am signed into my email provider all the time - clicking on a link in that email account is near-instant, and I don't need to remember yet another string of meaningless digits; Persona's new 'and you need a Persona password too' makes Persona as uninteresting as every other 'you can make an account here' system.

**Comment by shane-tomlinson** [comment]
@rbtcollins - Thanks for the report. 

Persona does _not_ need a password if all of the emails you use support the [Persona Identity Provider protocol](https://developer.mozilla.org/en/BrowserID/Guide_to_Implementing_a_Persona_IdP). Instead of asking for a Persona password, Persona will redirect you to your email provider, where you will authenticate with them.

Persona is based on the BrowserID protocol, which itself was based on the Verified Email Protocol. The basic premise is that Persona gives the site you are signing in to an email address that has been verified to be under the control of the user, at the time of address registration. If your email provider does not support the Persona Identity Provider protocol, Mozilla stands in to verify that you own the address. Mozilla is a "fallback" identity provider, and will vouch for your identity only so long as your email provider does not support the Persona Identity Provider protocol. As soon as your email provider is Persona ID Provider enabled, Mozilla stands down and no longer acts as your identity provider.

As part of the fallback Identity Provider, we ask you to create a fallback identity provider password. We ask you to create this password so that we do not have to ask you to verify that you own the email address every time you try to sign in to a site. Instead of asking you to verify every single time you sign in, we ask for your fallback identity provider password. This password is useless when your email providers support the Identity Provider protocol.

You can see how this works, today, using a Yahoo! address, or using our example Identity Provider, eyedee.me

Steps:
1) Open http://123done.org, click on the "Sign in With Persona"
2) If signed in to Persona, sign out.
3) In the email address field, type <some_username>@eyedee.me
4) You will see a screen that says "eyedee.me makes this easy!" and then you can click the button to redirect to eyedee.me
5) Enter in a new password at eyedee.me
6) You will return to Persona, the dialog will close, and you will be signed in to 123done.

Persona never asked you for a password in this flow.

The solution to your report is to ask your email provider to become a Persona Identity Provider.

**Comment by lloyd** [comment]
Which email provider do you use?

rbtcollins notifications@github.com wrote:

I am signed into my email provider all the time - clicking on a link in that email account is near-instant, and I don't need to remember yet another string of meaningless digits; Persona's new 'and you need a Persona password too' makes Persona as uninteresting as every other 'you can make an account here' system.

---

Reply to this email directly or view it on GitHub:
https://github.com/mozilla/browserid/issues/3494

**Comment by rbtcollins** [comment]
@shane-tomlinson Thanks for the detailed explanation- it is consistent with my existing understanding. What I would like is to do an email verification everytime with the fallback provider, because that is easier. If you kept a cookie live on my browser for say a month, and did an email verification that would be nice too, but that's obviously orthogonal to the question of whether you id my by my still having the email address, or by my remembering a password I created once, when asked for it by Persona, and have not thought about for the last $period of time. At the moment, my login for Persona sites goes like this:
'Login with persona'
-> email address
-> I forgot my password
-> click on link in email
-> grab a uuid and paste it into both fields

@lloyd Google Apps for Domains at the moment. So I have a browser tab open, and it's a moments work to switch to the tab and click on the 'yes I have this email' field. I think keeping that workflow + a cookie for some tasteful amount of caching, would give an excellent experience.

**Comment by lloyd** [comment]
We will release identity bridging for gmail, and I believe we will also be able to support Google apps for domains with this realease.  I hope this will give you the experience you expect :)

rbtcollins notifications@github.com wrote:

@shane-tomlinson Thanks for the detailed explanation- it is consistent with my existing understanding. What I would like is to do an email verification everytime with the fallback provider, because that is easier. If you kept a cookie live on my browser for say a month, and did an email verification that would be nice too, but that's obviously orthogonal to the question of whether you id my by my still having the email address, or by my remembering a password I created once, when asked for it by Persona, and have not thought about for the last $period of time. At the moment, my login for Persona sites goes like this:
'Login with persona'
-> email address
-> I forgot my password
-> click on link in email
-> grab a uuid and paste it into both fields

@lloyd Google Apps for Domains at the moment. So I have a browser tab open, and it's a moments work to switch to the tab and click on the 'yes I have this email' field. I think keeping that workflow + a cookie for some tasteful amount of caching, would give an excellent experience.

---

**Comment by callahad** [comment]
> I believe we will also be able to support Google apps for domains with this realease.

nope nope nope -- Google Apps for your Domain is a huge can of worms that I'm not going to open until we get Gmail and Hotmail online. ;)

**Comment by fmarier** [comment]
So to summarize/paraphrase @rbtcollins suggestion:

> The UX would be better if we made optional the fallback IdP's optimization of storing the confirmation state of email addresses.

Because that's really the only reason why we have a password and he's saying that remembering/resetting a password is more work than clicking on an email link, therefore it's the opposite of an optimization at all for him.

**Comment by lloyd** [comment]
Where can I read about what makes it so tricky?

Dan Callahan notifications@github.com wrote:

> I believe we will also be able to support Google apps for domains with this realease.

nope nope nope -- Google Apps for your Domain is a huge can of worms that I'm not going to open until we get Gmail and Hotmail online. ;)

---

**Comment by callahad** [comment]
@lloyd In short, GAFYD has Yahoo-style Aliases, dots are significant (gmail ignores them), and GAFYD accounts may or may not be directly attached to email accounts.

That last part is particularly weird. Consider Mozilla: we have GAFYD, but it's not connected directly to LDAP.  We don't use GAFYD's mail services. Yet I _can_ use my Mozilla/Google account over OpenID, despite that not proving anything about my ownership of an email address. :)

I suspect there are dozens of unknown, unanticipated corner cases here. Regardless, just the known ones (aliases especially)  are enough for me to say a strong "No" to the idea of exploring this before Yahoo and Gmail are fully resolved.

**Comment by callahad** [comment]
Basically, just because it's Google, doesn't mean we can make the same assumptions that we made with Gmail.

It's best to consider this as comparable in scale to building an entirely new bridge from scratch, with at least all of the same pitfalls of Yahoo, and possibly more.

**Comment by fmarier** [comment]
It sounds like GFYD would need a lot more testing to avoid a repeat of the Yahoo problems. I'm with @callahad for keeping it out of Beta3.

**Comment by lloyd** [comment]
ok, because all of this is tracked elsewhere, I'm closing down this bug, rest assured GFYD is a desired future feature.

**Comment by rbtcollins** [comment]
Sorry, I am confused. Where is the 'you have to have a password with the fallback provider' tracked?

**Comment by seanmonstar** [comment]
Indeed, they did a terrific job of hijacking this issue, huh? I'm thinking we just re-open this, and delete the comments about GAFYD, since that conversation has happened plenty of times elsewhere.

**Comment by fmarier** [comment]
Updating the bug title...

**Comment by lloyd** [comment]
I'm confused.  What is this bug actually tracking?  Are we going to remove the password from the fallback IdP?  How?

**Comment by jaredhirsch** [comment]
I don't think it's worthwhile to leave this bug open as it stands. There's nothing actionable here--we're planning to run the fallback _forever_.

We generally confine speculative feature discussion to the public list, IMO that's where this conversation belongs.

**Comment by callahad** [comment]
The concrete question / proposal is: Should the fallback require a password, or should it rely solely on session cookies?

**Comment by jaredhirsch** [comment]
So this is reopening the ["why does persona have a password" thread](https://groups.google.com/d/topic/mozilla.dev.identity/p2ewIlr5ypo/discussion) from April, then. Can we shut down this bug and move the discussion to the list?

**Comment by jaredhirsch** [comment]
Closing this bug. If you have concerns about the fallback requiring a password, please take it to the list.

**Comment by rbtcollins** [comment]
I opened this bug when @fmarier asked me to after I whinged on twitter; if you consider it a feature, thats fine, just say so. If you aren't decided about it, closing the issue seems like entirely the wrong thing to do. That said, I'm not engaged enough with Persona development to join the mailing list, so no, I won't take it there. You have the information now [that I [and btw others that have complained about this in various locations]] don't see the point of *yet another password proffering scheme', which Persona is advertised as not-being, but with this anti-feature, has become one.
